File: totoggle_go/internal/strategy/percentage.go
package strategy

import (
	"hash/fnv"
	"math/rand"
	"strconv"

	"github.com/manorfm/toToggles/totoggle_go/internal/toggle"
)

// PercentageEvaluator activates a configured percentage of evaluations. With a key (a stable
// per-user/session identifier), the bucket is deterministic — the same key + rule value always
// lands in the same bucket, via FNV-1a (Go has no equivalent to Java's JLS-specified
// String.hashCode(), so this deliberately isn't bit-identical to the Kotlin client's bucketing —
// only self-consistent within one client, which is what "same user always gets the same result"
// actually requires). With no key, there is nothing to be consistent with, so it falls back to a
// per-call random draw.
type PercentageEvaluator struct {
	randomBucket func() float64 // returns a value in [0, 1)
}

// NewPercentageEvaluator builds a PercentageEvaluator. randomSource, if non-nil, is used as the
// source of randomness for no-key evaluations (returning a float in [0, 1)) — tests inject a
// fixed source instead of depending on real randomness. A nil randomSource defaults to
// math/rand.
func NewPercentageEvaluator(randomSource func() float64) PercentageEvaluator {
	if randomSource == nil {
		randomSource = rand.Float64
	}
	return PercentageEvaluator{randomBucket: randomSource}
}

func (p PercentageEvaluator) Evaluate(rule toggle.ActivationRule, key string, hasKey bool) bool {
	percentage, err := strconv.ParseFloat(rule.Value, 64)
	if err != nil {
		return false
	}
	if percentage < 0 || percentage > 100 {
		return false
	}

	var bucket float64
	if hasKey {
		bucket = consistentBucket(rule.Value, key)
	} else {
		bucket = p.randomBucket() * 100
	}
	return bucket < percentage
}

// consistentBucket derives a deterministic value in [0, 100) from ruleValue+key.
func consistentBucket(ruleValue, key string) float64 {
	h := fnv.New32a()
	_, _ = h.Write([]byte(ruleValue + ":" + key))
	return float64(h.Sum32()%10000) / 100.0
}

File: totoggle_node/src/internal/strategy/percentage.ts
import type { ActivationRule } from "../toggle/rule.js";
import type { Evaluator } from "./strategy.js";

const FNV_OFFSET_BASIS = 0x811c9dc5;
const FNV_PRIME = 0x01000193;

/** FNV-1a, 32-bit. Neither Node nor a zero-dependency environment exposes anything like Java's
 * JLS-specified String.hashCode(), so this is hand-written — a few lines, no dependency. */
function fnv1a(input: string): number {
  let hash = FNV_OFFSET_BASIS;
  // UTF-8 bytes make the result identical to the Go and JVM implementations.
  for (const byte of new TextEncoder().encode(input)) {
    hash ^= byte;
    hash = Math.imul(hash, FNV_PRIME);
  }
  return hash >>> 0;
}

/** A deterministic value in [0, 100) derived from ruleValue+key. */
function consistentBucket(ruleValue: string, key: string): number {
  const hash = fnv1a(`${ruleValue}:${key}`);
  return (hash % 10000) / 100;
}

const PLAIN_DECIMAL = /^-?\d+(\.\d+)?$/;

function parsePercentage(raw: string): number | null {
  const trimmed = raw.trim();
  if (!PLAIN_DECIMAL.test(trimmed)) {
    return null;
  }
  return Number(trimmed);
}

/**
 * Activates a configured percentage of evaluations. With a key (a stable per-user/session
 * identifier), the bucket is deterministic — the same key + rule value always lands in the same
 * bucket. This is deliberately not bit-identical to totoggle_java's or totoggle_go's bucketing
 * (neither Java's String.hashCode() nor Go's FNV-1a-via-hash/fnv is replicated bit-for-bit here)
 * — now deliberately bit-identical across ToToggle SDKs. The supplied key includes the toggle
 * path, so two flags with the same percentage do not share an accidental cohort. Without a
 * rollout key, callers fail closed before reaching this evaluator.
 */
export class PercentageEvaluator implements Evaluator {
  constructor(private readonly randomSource: () => number = Math.random) {}

  evaluate(rule: ActivationRule, key: string | undefined): boolean {
    const percentage = parsePercentage(rule.value);
    if (percentage === null || percentage < 0 || percentage > 100) {
      return false;
    }

    const bucket = key !== undefined ? consistentBucket(rule.value, key) : this.randomSource() * 100;
    return bucket < percentage;
  }
}
